File: app/api/portfolio.py
"""
Portfolio API Router
====================

Funds, Positions, Trades, OrderBook endpoints.

Author: Antony HFT System
"""

from fastapi import APIRouter, HTTPException, Query
from app.services.portfolio_service import PortfolioService

router = APIRouter(prefix="/portfolio", tags=["Portfolio"])


# ═══════════════════════════════════════════════════════════════════════════════
# FUNDS & MARGIN
# ═══════════════════════════════════════════════════════════════════════════════

@router.get("/funds")
async def get_funds(segment: str = Query("SEC", description="SEC=Equity, COM=Commodity")):
    """
    Get Funds and Margin
    
    Returns available margin, used margin for order placement validation.
    
    Response:
    ```json
    {
        "available_margin": 50000,
        "used_margin": 10000,
        "payin_amount": 0
    }
    ```
    """
    try:
        return await PortfolioService.get_funds_and_margin(segment)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ═══════════════════════════════════════════════════════════════════════════════
# POSITIONS
# ═══════════════════════════════════════════════════════════════════════════════

@router.get("/positions")
async def get_positions():
    """
    Get Intraday Positions
    
    Returns all open positions with real-time P&L.
    
    Response:
    ```json
    {
        "data": [
            {
                "instrument_token": "NSE_FO|61755",
                "quantity": 25,
                "average_price": 200,
                "ltp": 210,
                "pnl": 250
            }
        ]
    }
    ```
    """
    try:
        return await PortfolioService.get_positions()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/holdings")
async def get_holdings():
    """
    Get Long-term Holdings (Delivery)
    
    Returns all delivery holdings.
    """
    try:
        return await PortfolioService.get_holdings()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ═══════════════════════════════════════════════════════════════════════════════
# TRADES
# ═══════════════════════════════════════════════════════════════════════════════

@router.get("/trades")
async def get_trades():
    """
    Get Today's Trades
    
    Returns all executed trades for the day.
    """
    try:
        return await PortfolioService.get_trades()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/trades/{order_id}")
async def get_trades_by_order(order_id: str):
    """
    Get Trades for specific Order
    """
    try:
        return await PortfolioService.get_trades_by_order(order_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ═══════════════════════════════════════════════════════════════════════════════
# ORDER BOOK
# ═══════════════════════════════════════════════════════════════════════════════

@router.get("/orders")
async def get_order_book():
    """
    Get Order Book
    
    Returns all orders - pending, complete, rejected.
    """
    try:
        return await PortfolioService.get_order_book()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/orders/{order_id}")
async def get_order_details(order_id: str):
    """
    Get Order Details
    """
    try:
        return await PortfolioService.get_order_details(order_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/orders/{order_id}/history")
async def get_order_history(order_id: str):
    """
    Get Order History (status changes)
    """
    try:
        return await PortfolioService.get_order_history(order_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/lib/stores/portfolio.ts
// Portfolio Store - Funds, Positions, Holdings
const API_BASE = 'http://localhost:8000';

// Get Funds
export async function getFunds(segment: 'SEC' | 'COM' = 'SEC') {
    try {
        const res = await fetch(`${API_BASE}/api/v1/portfolio/funds?segment=${segment}`);
        return await res.json();
    } catch (err) {
        console.error('getFunds failed:', err);
        return null;
    }
}

// Get Positions
export async function getPositions() {
    try {
        const res = await fetch(`${API_BASE}/api/v1/portfolio/positions`);
        return await res.json();
    } catch (err) {
        console.error('getPositions failed:', err);
        return null;
    }
}

// Get Holdings
export async function getHoldings() {
    try {
        const res = await fetch(`${API_BASE}/api/v1/portfolio/holdings`);
        return await res.json();
    } catch (err) {
        console.error('getHoldings failed:', err);
        return null;
    }
}

// Get Trades
export async function getTrades() {
    try {
        const res = await fetch(`${API_BASE}/api/v1/portfolio/trades`);
        return await res.json();
    } catch (err) {
        console.error('getTrades failed:', err);
        return null;
    }
}

// Get Order Book
export async function getOrderBook() {
    try {
        const res = await fetch(`${API_BASE}/api/v1/portfolio/orders`);
        return await res.json();
    } catch (err) {
        console.error('getOrderBook failed:', err);
        return null;
    }
}

// Get Order Details
export async function getOrderDetails(orderId: string) {
    try {
        const res = await fetch(`${API_BASE}/api/v1/portfolio/orders/${orderId}`);
        return await res.json();
    } catch (err) {
        console.error('getOrderDetails failed:', err);
        return null;
    }
}
